### ISSUE TITLE
Add more attributes to events

### ISSUE THREAD

**Issue by cortadocodes** [octue/octue-sdk-python#634]
### Use Case
I'd like to:
- [ ] Add a `from` or `sender_sruid` attributes to make it easier to find the SRUID of the child from a message
- [ ] Rename `message_number` to `event_number` for consistency

**Comment by thclark** [comment]
What about a no-store flag which if true (default false) would prevent things from being streamed to any long term storage? Essentially this would allow the originator to request that the chain of events isn’t stored anywhere - great for use cases like testing against live services.

**Comment by thclark** [comment]
in terms of the ID didn't we establish that messages have an ID already?

**Comment by thclark** [comment]
Also should the sender be split into ~~`sender-namespace`~~, `sender-name`, `sender-version` allowing refined filtering to a whole service or just a revision.

(This has made me think that the namespace is probably common to the whole workspace, making it redundant for this purpose)

**Comment by cortadocodes** [comment]
> What about a no-store flag which if true

Good idea. There's a slightly related `forward_logs` attribute too that's maybe a bit outdated. Maybe we should remove that and add `no_store`

> in terms of the ID didn't we establish that messages have an ID already?

There is but it's pub/sub specific so won't be there for other backends and won't be consistent with the ID we add for them.

> Also should the sender be split into sender-namespace, sender-name, sender-version allowing refined filtering to a whole service or just a revision.

I think we should add the namespace as we may have more than one namespace in a workspace. What's the use case for splitting the SRUID into its three parts instead of keeping it whole?

**Comment by cortadocodes** [comment]
How would the `no_store` attribute relate with the `save_diagnostics` attribute? Does the latter respect the former?

**Comment by thclark** [comment]
> > What about a no-store flag which if true
> 
> Good idea. There's a slightly related `forward_logs` attribute too that's maybe a bit outdated. Maybe we should remove that and add `no_store`

Now that we have a store, if we wanted **not** to store logs in general, we would probably store them but clean up after 7 days (or whatever period).

Of course if you don't want to see logs, but don't care about service performance (or maybe want them stored for posterity), you can still stream them to the store and simply filter out the log event kind in your subscription or query.

However, the `forward_logs` attribute would still cover the narrow use case of accelerating communication (by allowing you to avoid incurring the expense of sending them TO pubsub in the first place). I've no idea how valuable this is; but it's certainly something we've never actually used.

I would leave it alone, since it's still a legitimate feature albeit unused to date... **unless** it's getting in the way somehow in which case feel free to steamroller it.

> > in terms of the ID didn't we establish that messages have an ID already?
> There is but it's pub/sub specific so won't be there for other backends and won't be consistent with the ID we add for them.

ah, ok

 
> > Also should the sender be split into sender-namespace, sender-name, sender-version allowing refined filtering to a whole service or just a revision.
> I think we should add the namespace as we may have more than one namespace in a workspace. 
I'd never intended that to be a possibility. Surely if you were doing services all in one project/workspace they'd be namespaced together. That said, leaving the namespace there would keep open the option for later (also keep openthe option for doing things like bridging across tables in different namespaces/workspaces to merge events from multiple services... so I'm fine wiht keeping it.

What's the use case for splitting the SRUID into its three parts instead of keeping it whole?

You want to be able to filter on:
  - all events from a namespace
  - all events from a service
  - all events from a service revision
 
 My understanding of the filters is that they use equality match (`if attribute == value`).
**If you have an attribute `sruid`:**
 - you get service revision events by doing `if sruid ==  'octue/ourservice/0.1.9'` 
 - but how would you get all events from `octue/ourservice`? 
 
**Whereas if you had attributes `namespace` `service` `version`:**
- you get service revision events with `if namespace=='octue' && service == 'ourservice' && version == '0.1.9`
- you get all service events with `if namespace=='octue' && service == 'ourservice'`

**Comment by cortadocodes** [comment]
> but how would you get all events from octue/ourservice?

You could actually do this with the filter `hasPrefix(attributes.sruid, "octue/ourservice:")` - see [here](https://cloud.google.com/pubsub/docs/subscription-message-filter#filtering_syntax) but it wouldn't work if you wanted, for example, all services with the name `ourservice` on any namespace.

If we're splitting the SRUID like this, I'll have to do it for three fields, not just one:
- Originator
- Sender
- Recipient

Is that ok or too much?

**Comment by thclark** [comment]
That's actually too much; if there's a hasPrefix filter that totally meets our use case without overcompensating things.

(I thought you could only filter on match)

I can't imagine a clear use case where we'd want the later example you have (all services named x from different namespaces) so let's not worry about that.

(TBH I'm still not convinced that namespace would ever change at all in these tables anyhow, rendering it moot)

**Comment by cortadocodes** [comment]
Oh no I just added and tested all of those

**Comment by cortadocodes** [comment]
> How would the no_store attribute relate with the `save_diagnostics` attribute? Does the latter respect the former or are they independent?

Still wondering about this. 

**Save diagnostics**
When `save_diagnostics=True` for a question asked to a child, these things are uploaded to a cloud storage bucket for each question:
- Configuration values
- Configuration manifest and datasets
- Input values
- Input manifests and datasets
- The events received from any children the child asked its own questions to as part of the question

When `save_diagnostics=False`, none of this is uploaded.

**Store events**
In contrast, when `no_store=False`, these things are saved in the event store:
- The question event, which includes:
  - Input values
  - Input manifest (which can be used to get the input datasets)
- All events sent by the child back to the parent, including the result event which includes:
  - Output values
  - Output manifest (which can be used to get the output datasets)

When `no_store=True`, none of these will be stored.
